### DISCUSSION TITLE
Confusion About Access Tokens

### DISCUSSION THREAD

**Discussion by kevinswarner** [nextauthjs/next-auth#2000] [Help]
I have read through the documentation and many of the issues, but I am still a little confused about tokens.

We were previously using a client-side authentication approach to our provider, and were receiving back three tokens (id token, access token, and refresh token). Our client side app then included the access token in an authorization header in all calls to our external API so that our API could verify the identity of the user making the API call.

We have had some concerns about storing the access token from the provider in the client, so we are moving to NextAuth in hopes of mitigating our security concerns.

My confusion is about the provider tokens vs NextAuth tokens. I now realize they are actually different.

My main question is this...

**Is it safe and appropriate to use the NextAuth token when making calls to our external API, as opposed to the provider access token?**

We are planning to proxy the calls to our external API through a NextJS API route. In that route, we are planning on grabbing the token using jwt.getToken() and then adding that to the header in the request to the external API. I originally was thinking I needed to get the original provider token and use that, but that is seeming overly complicated. That is why I am wondering if I should really just use the token that NextAuth creates, rather than the provider token.

If I made that change, I am assuming in our external API we would simply need to use the same "secret" to verify the token as I have specified in my NextAuth config.

I guess my confusion if related to the difference between the NextAuth session vs the provider session. I had been thinking these should really be the same, but now I am thinking that they are separate. What I mean is that whether the user is logged in or logged out of the provider is separate from whether the user is logged in or logged out of our app using NextAuth, except for the fact that the NextAuth session is created based on whether or not the user is logged into the provider.

If anyone has any thoughts that might help me understand this, I would greatly appreciate it!

**Comment by harleyharl** [comment]
I'm having the same issue at the moment. I'm going from my own implementation of client-side authentication to server-side authentication with Next-auth. I'm using a custom OAuth provider and my login flow works properly using NextAuth, but the tokens saved in the session are not the access tokens returned from my OAuth provider. I guess they are an encrypted version of my access tokens. So the access tokens I actually need to persist to use in making any API calls aren't available from the session cookie object. 

Can anyone please point me in the right direction? Should I be using the SignIn or Jwt callback to save my access token to local storage or in the cookie?
